Proposed-RayQueryAndFriends

MakoEnergy opened the pull request:
This contains misc bug fixes for things I encountered along the way of working on other tasks.  It also contains some initial code for the Player system.  The bulk of the code is centered around the RayQuery interface class and it's implementations in the Physics and Graphics subsystems.

Sqeaky commented:
This is a general issue, and not something we shoulf fix with this PR, but the entresol API needs to be reworked so that the GetManager method doesn't require static_casts, that is gross.

MakoEnergy commented:
We used to have "GetSceneManager()" methods for each manager class before.  However I removed them to prepare for the decoupling of code we'll get in the Jagati.  We could do something similar to what I added to the RenderableProxy and CollidableProxy classes with the "_Upcast(*)" methods.  

Such as have a "SceneManager::Upcast(WorldManager* Man)" or "InputManager::Upcast(EntresolManager* Man)" methods.  Could use a static_cast or dynamic_cast for the implementation.  That would be a little bit cleaner and keep things decoupled.

Sqeaky commented:
I agree postponing that decision for the Jagati refactor is a good idea.

Sqeaky commented:
The initialize function on the MousePicker class should accept pointers with a concept of ownership. It is impossible to tell from just the definition if those pointers are owned or not be the MousePicker. The deinitialize function compounds this by making it impossible if it is or isn't own until this is called with a potentially dynamic.

We should follow Bjarnes Rules on this ( https://github.com/isocpp/CppCoreGuidelines )

* A naked pointer in the API should not imply ownership.
* Passing of owned pointers should be done by some class that specifies lifetime like unique_ptr or shared_ptr.

Sqeaky commented:
There are several instances of remarked out code, like the MeshRawBufferHelper, consider removing or stashing it not using.

Sqeaky commented on Mezzanine/src/MathTools/intersections.cpp:
```diff
@@ -56,6 +58,9 @@ namespace
     typedef std::pair<Mezzanine::Real,Mezzanine::Real> SegmentPosPair;
+    /// @brief Convenience epsilon value used for Plane checks.
+    /// @remarks It's purposefully less precise than what is returned by std::numeric_limits in order to add more stability to checks.
+    const Mezzanine::Real PlaneCheckEpsilon = 1e-6f;
```
Why not use to the type_traits::<Mezzanine::Real>?

MakoEnergy replied:
I think the algorithm needs more leeway than what an actual epsilon value would provide.  Even several times the epsilon.  I suppose that makes the name "Epsilon on here a bit of a misnomer, but that is how it was referenced in the academic paper I read to implement the algorithm.

MakoEnergy commented:
The MeshRawBufferHelper in submesh.h was an experiment that was taking too long to complete.  I may come back to it.  

I've also already pushed a commit that removes any notion of ownership from the MousePicker.

Sqeaky commented:
What is the magic significance of 15000.0? 

in Mezzanine/src/Physics/collidablerayquery.cpp.

Sqeaky commented:
"process" is the Worst Method name ever, (except for "function", "method" and "do")!

Bullet made you do this right?

also in Mezzanine/src/Physics/collidablerayquery.cpp

Sqeaky commented:
Why is 30 proxies better than 10?

In Mezzanine/src/Physics/managerconstructioninfo.h

Sqeaky commented:
On the LineSegment2D in Mezzanine/src/linesegment.h, because you have a copy constructor and assignment operators you implicitly delete the move operator and move assignment. It looks like all 4 could be defaulted.

MakoEnergy commented:
1) It's arbitrary, however the number needs to be there and needs to be high because Bullet does it's ray casts not with rays, but with line segments.
2) Yes, that is a bullet limitation.
3) Because 10 was crashing the unit tests I made.

Sqeaky commented:
1. It should parameterized then, because we don't know what statically what a sensible number is.

2. That sucks.

Sqeaky commented:
Similar issues exist on the LineSegment3D as the LineSegment2D in terms of move semantic implicit deletion.

Sqeaky commented:
The player stuff looks incomplete, its not supposed to be part of this PR right?

MakoEnergy commented:
I'll have the LineSegment stuff and the CollidableRayQuery distance parameter in an upcoming commit.  

The player stuff is boilerplate for code I was working on prior to the RayQuery.  It's a WIP and not the focus on the PR.

Sqeaky commented:
In rayquery.h the constructor for the RayQueryHist does not set 3 of the members. This is most trouble for WorldProxy* because it makes it easier to write potentially pernicious memory corruption bugs.

Sqeaky commented:
While dealing with copy and move semantics, you might want to look at the triangle class.

Sqeaky commented:
in the unit tests you #include "sstream" this is probably just a typo, but will make builds fail on some platforms unless we are provide an sstream implmentation.

MakoEnergy commented:
Added a couple more commits that are the requested changes...and a little extra.
